### PR TITLE
Update tokio-udp to use std-future

### LINKED ISSUE #616: Async/Await interface for UDP
## Version

```
├── tokio v0.1.8                                                                
│   ├── tokio-codec v0.1.0
│   │   └── tokio-io v0.1.8
│   ├── tokio-current-thread v0.1.1
│   │   └── tokio-executor v0.1.4
│   ├── tokio-executor v0.1.4 (*)
│   ├── tokio-fs v0.1.3
│   │   ├── tokio-io v0.1.8 (*)
│   │   └── tokio-threadpool v0.1.6
│   │       └── tokio-executor v0.1.4 (*)
│   │   └── tokio-io v0.1.8 (*)
│   ├── tokio-io v0.1.8 (*)
│   ├── tokio-reactor v0.1.5
│   │   ├── tokio-executor v0.1.4 (*)
│   │   └── tokio-io v0.1.8 (*)
│   ├── tokio-tcp v0.1.1
│   │   ├── tokio-io v0.1.8 (*)
│   │   └── tokio-reactor v0.1.5 (*)
│   ├── tokio-threadpool v0.1.6 (*)
│   ├── tokio-timer v0.2.6
│   │   └── tokio-executor v0.1.4 (*)
│   ├── tokio-udp v0.1.2
│   │   ├── tokio-codec v0.1.0 (*)
│   │   ├── tokio-io v0.1.8 (*)
│   │   └── tokio-reactor v0.1.5 (*)
│   └── tokio-uds v0.2.1
│       ├── tokio-io v0.1.8 (*)
│       └── tokio-reactor v0.1.5 (*)
├── tokio-core v0.1.17
│   ├── tokio v0.1.8 (*)
│   ├── tokio-executor v0.1.4 (*)
│   ├── tokio-io v0.1.8 (*)
│   ├── tokio-reactor v0.1.5 (*)
│   └── tokio-timer v0.2.6 (*)
└── tokio-io v0.1.8 (*)
```

## Platform

Darwin xxxxxxx 17.7.0 Darwin Kernel Version 17.7.0: Thu Jun 21 22:53:14 PDT 2018; root:xnu-4570.71.2~1/RELEASE_X86_64 x86_64

## Description

I have code to join a UDP socket to a multicast group here:

[socket2 multicast join code](https://gist.github.com/pusateri/df98511b88e9000f388d344a1f3db9e7)

I want to take the received UDP multicast packets and iterate over them using the new async/await pattern that is now available for TCP with .incoming but this isn't available for UDP.

### LINKED ISSUE #1126: send_dgram doesn't work with connect on macOS
## Version
```
└── tokio v0.1.21
    ├── tokio-codec v0.1.1
    │   └── tokio-io v0.1.12
    ├── tokio-current-thread v0.1.6
    │   └── tokio-executor v0.1.7
    ├── tokio-executor v0.1.7 (*)
    ├── tokio-fs v0.1.6
    │   ├── tokio-io v0.1.12 (*)
    │   └── tokio-threadpool v0.1.14
    │       └── tokio-executor v0.1.7 (*)
    │   └── tokio-io v0.1.12 (*)
    ├── tokio-io v0.1.12 (*)
    ├── tokio-reactor v0.1.9
    │   ├── tokio-executor v0.1.7 (*)
    │   ├── tokio-io v0.1.12 (*)
    │   └── tokio-sync v0.1.5
    ├── tokio-sync v0.1.5 (*)
    ├── tokio-tcp v0.1.3
    │   ├── tokio-io v0.1.12 (*)
    │   └── tokio-reactor v0.1.9 (*)
    ├── tokio-threadpool v0.1.14 (*)
    ├── tokio-timer v0.2.11
    │   └── tokio-executor v0.1.7 (*)
    ├── tokio-trace-core v0.2.0
    ├── tokio-udp v0.1.3
    │   ├── tokio-codec v0.1.1 (*)
    │   ├── tokio-io v0.1.12 (*)
    │   └── tokio-reactor v0.1.9 (*)
    └── tokio-uds v0.2.5
        ├── tokio-codec v0.1.1 (*)
        ├── tokio-io v0.1.12 (*)
        └── tokio-reactor v0.1.9 (*)
```

## Platform
`Darwin computername.local 18.2.0 Darwin Kernel Version 18.2.0: Thu Dec 20 20:46:53 PST 2018; root:xnu-4903.241.1~1/RELEASE_X86_64 x86_64`

## Subcrates
tokio-udp

## Description
I'm creating a new socket with `let socket = tokio::net::UdpSocket::bind(&local_addr).unwrap()` and then limit communication to a specific client using `socket.connect(&client_addr).unwrap()`.
All documentation suggests that `recv` and `send` shall now be used instead of `recv_from` and `send_to`.
However, tokio::net::UdpSocket's `recv_dgram` and `send_dgram` functions always use `recv_from` and `send_to` internally. `send_dgram` also requires to specify a target IP address.

I can successfully use `send_dgram` under Windows and Linux by passing `client_addr` as the target IP address: `socket.send_dgram(buf, &client_addr)`. Their underlying `send_to` system API apparently works after `connect` when specifying the same client IP address.
However, the same code fails under macOS with `Socket is already connected (os error 56)`.
This seems to be a known limitation of macOS, see e.g. https://www.perlmonks.org/bare/?node_id=415228

Possible solutions I imagine:
- A new pair of functions `recv_dgram_connected` and `send_dgram_connected` returning futures for working on UDP sockets where `connect` has been called. They should call `recv` and `send` internally.
- Something like a `recv_dgram_filtered` function returning a future that only yields `Async::Ready` when a datagram has been received *and* it is from the given IP address (or based on a different filter criteria), thereby performing the filtering `connect` does at user level.

### LINKED ISSUE #1195: udp: update to std::future
`tokio-udp` should be updated to use `std::future`.

This issue is to track discussion and progress.

### PR DESCRIPTION
**Pull request by sopium**
Close #1195.

Changes:

* `recv_dgram`, `send_dgram`, `RecvDgram` and `SendDgram` are replaced with `send`, `recv`,
  `send_to` and `recv_from` which return std `Future`s that borrows the socket, i.e., can be used with async/await. It's roughly the same API that runtime and romio offers. (Close #616, Close #1126.)

* The `poll...` functions are updated.

* `UdpFramed` is removed (for now). It seems that `tokio-codec` hasn't been updated yet. Besides, I think the framed API is not as useful as it used to be in v0.1:
  + `UdpFramed` produces/consumes owned buffers, which is the preferred way to deal with buffers when using combinators. But with async/await borrowed buffers can be used just fine.
  + `UdpFramed` can be split. But it incurs an unnecessary locking overhead. A specialized split should be implemented (like #174).

* Docs are updated.

* Some basic tests are included. I have fixed and exposed the `tokio::test` macro and let `tokio_udp` depend on `tokio` for testing. Feels quite hacky. Let me know if there are better ways.

### CODE REVIEW

File: tokio-macros/src/lib.rs [hunk 1]
```diff
@@ -79,7 +79,7 @@ pub fn test(_attr: TokenStream, item: TokenStream) -> TokenStream {
         #(#attrs)*
         fn #name() #ret {
             let mut rt = tokio::runtime::current_thread::Runtime::new().unwrap();
-            rt.block_on_async(async { #body })
+            rt.block_on(async { #body })
```
Comment (LucioFranco) [inline tokio-macros/src/lib.rs]: change seems unrelated?

File: tokio-udp/src/lib.rs [hunk 2]
```diff
@@ -10,23 +10,28 @@
...
-//! [`RecvDgram`]: struct.RecvDgram.html
-//! [`SendDgram`]: struct.SendDgram.html
-//! [`UdpFramed`]: struct.UdpFramed.html
-//! [`framed`]: struct.UdpSocket.html#method.framed
+//! [`Recv`], [`Send`], [`RecvFrom`] and [`SendTo`] structs respectively.
 
-mod frame;
-mod recv_dgram;
-mod send_dgram;
+macro_rules! ready {
```
Comment (LucioFranco) [inline tokio-udp/src/lib.rs]: this should be imported from `tokio-futures`

File: tokio-udp/src/recv.rs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+use std::io;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+/// A future that receives a datagram from the connected address.
+///
+/// This `struct` is created by [`recv`](super::UdpSocket::recv).
+#[must_use = "futures do nothing unless polled"]
+#[derive(Debug)]
+pub struct Recv<'socket, 'buf> {
```
Comment (LucioFranco) [inline tokio-udp/src/recv.rs]: just to be consistent, Ive been just using `'a, 'b` instead of naming them.

File: tokio-udp/src/socket.rs [hunk 4]
```diff
@@ -52,13 +53,19 @@ impl UdpSocket {
...
-        }
+    /// Returns a future that sends data on the socket to the remote address to which it is connected.
+    /// On success, the future will resolve to the number of bytes written.
+    ///
+    /// The [`connect`] method will connect this socket to a remote address. The future
+    /// will resolve to an error if the socket is not connected.
+    ///
+    /// [`connect`]: #method.connect
+    ///
+    /// # Panics
```
Comment (LucioFranco) [inline tokio-udp/src/socket.rs]: I think we can drop this, and this should be a detail on the future docs it self. Most users will be using this within an async context anyways. This was a much bigger issue with futures 0.1.

File: tokio-udp/src/socket.rs [hunk 5]
```diff
@@ -52,13 +53,19 @@ impl UdpSocket {
...
+    ///
+    /// If the future is polled outside of a task context.
+    pub fn send<'socket, 'buf>(&'socket mut self, buf: &'buf [u8]) -> Send<'socket, 'buf> {
```
Comment (LucioFranco) [inline tokio-udp/src/socket.rs]: I believe you can elide the 'socket lifetime?

File: tokio-udp/src/socket.rs [hunk 6]
```diff
@@ -70,35 +77,49 @@ impl UdpSocket {
...
     }
 
-    #[deprecated(since = "0.1.2", note = "use poll_recv instead")]
-    #[doc(hidden)]
-    pub fn recv(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-        match self.poll_recv(buf)? {
-            Async::Ready(n) => Ok(n),
-            Async::NotReady => Err(io::ErrorKind::WouldBlock.into()),
-        }
+    /// Returns a future that receives a single datagram message on the socket from
```
Comment (LucioFranco) [inline tokio-udp/src/socket.rs]: these docs seem wrong for recv, they seem like they are for recv_from?

File: tokio-udp/src/socket.rs [hunk 7]
```diff
@@ -115,35 +136,46 @@ impl UdpSocket {
...
+    /// On success, returns `Poll::Ready(Ok(num_bytes_read))`.
     ///
     /// If no data is available for reading, the method returns
-    /// `Ok(Async::NotReady)` and arranges for the current task to receive a
+    /// `Poll::Pending` and arranges for the current task to receive a
     /// notification when the socket becomes receivable or is closed.
     ///
     /// # Panics
     ///
     /// This function will panic if called from outside of a task context.
```
Comment (LucioFranco) [inline tokio-udp/src/socket.rs]: this isnt the case anymore

File: tokio-udp/src/socket.rs [hunk 8]
```diff
@@ -115,35 +136,46 @@ impl UdpSocket {
...
-            Async::Ready(n) => Ok(n),
-            Async::NotReady => Err(io::ErrorKind::WouldBlock.into()),
-        }
+    /// Returns a future that sends data on the socket to the given address.
+    /// On success, the future will resolve to the number of bytes written.
+    ///
+    /// The future will resolve to an error if the IP version of the socket does
+    /// not match that of `target`.
+    ///
+    /// # Panics
```
Comment (LucioFranco) [inline tokio-udp/src/socket.rs]: again not the case anymore

File: tokio-udp/src/socket.rs [hunk 9]
```diff
@@ -213,74 +236,61 @@ impl UdpSocket {
...
     /// The socket will remain in a read-ready state until calls to `poll_recv`
-    /// return `NotReady`.
+    /// return `Poll::Pending`.
     ///
     /// # Panics
     ///
     /// This function panics if:
     ///
     /// * `ready` includes writable.
     /// * called from outside of a task context.
```
Comment (LucioFranco) [inline tokio-udp/src/socket.rs]: not the case anymore

File: tokio-udp/src/lib.rs (hunk 2 above)
Comment (carllerche) [reply tokio-udp/src/lib.rs]: This is TBD. Right now we have just been copy / pasting it around... I would just say keep it for now and we can fix later once we pick a strategy.

Comment (LucioFranco) [reply tokio-udp/src/lib.rs]: Yeah, though I kinda wanted to avoid copying it everywhere 🤷‍♂ but im fine with keeping it.

File: tokio-macros/src/lib.rs (hunk 1 above)
Comment (sopium) [reply tokio-macros/src/lib.rs]: Yeah, it should probably be a separate PR (like #1198).

File: tokio-udp/src/socket.rs (hunk 6 above)
Comment (sopium) [reply tokio-udp/src/socket.rs]: What's wrong? I just rephrased the doc of `poll_recv`, which is the same as `std::net::UdpSocket::recv`.

Comment (LucioFranco) [reply tokio-udp/src/socket.rs]: Ahhhh I confused RecvFrom and Recv sorry 😄

File: tokio-udp/src/recv.rs (hunk 3 above)
Comment (sopium) [reply tokio-udp/src/recv.rs]: I used `<'socket, 'buf>` because that's what runtime uses. But `SendTo` have a third field `target`, and having `<'socket, 'buf, 'target>` seems a bit ugly, so I used `<'socket, 'b>` which is not consistent itself.

I think I will switch to just `<'a, 'b>`.

File: tokio-udp/src/socket.rs (hunk 5 above)
Comment (sopium) [reply tokio-udp/src/socket.rs]: Yes, it's possible, but mark it explicitly seems clearer.

Comment (LucioFranco) [reply tokio-udp/src/socket.rs]: Not always true, generally speaking anytime you have a `fn(&self)` you know there is an anon lifetime so it doesnt make much sense to include it as it just adds noise.

Comment (sopium) [reply tokio-udp/src/socket.rs]: It seems very strange to elide only one lifetime. It breaks symmetry. And I don't see code out there do this:

* https://rust-lang-nursery.github.io/futures-api-docs/0.3.0-alpha.16/futures/io/trait.AsyncReadExt.html?search=#method.copy_into
* https://docs.rs/runtime/0.3.0-alpha.6/runtime/net/struct.UdpSocket.html#method.send_to
* https://docs.rs/romio/0.3.0-alpha.9/romio/udp/struct.UdpSocket.html#method.send_to

File: tokio-udp/src/lib.rs (hunk 2 above)
Comment (carllerche) [reply tokio-udp/src/lib.rs]: I mostly mean, uncopying can be handled in a follow up PR that does so across the repo.

File: tokio-macros/src/lib.rs (hunk 1 above)
Comment (carllerche) [reply tokio-macros/src/lib.rs]: I'm ok bundling this in the PR.

**Comment by carllerche** [comment]
Thanks

**Comment by carllerche** [comment]
Just need @LucioFranco to +1 as well.
